src/Catalyst.Core.Lib.Tests/UnitTests/IO/Messaging/Correlation/MessageCorrelationManagerTests.cs

```diff
@@ -0,0 +1,232 @@
+            cacheEntry.ExpirationTokens.Returns(expirationTokens);
+            var expirationCallbacks = new List<PostEvictionCallbackRegistration>();
+            cacheEntry.PostEvictionCallbacks.Returns(expirationCallbacks);
+
+            Cache.CreateEntry(correlationId).Returns(cacheEntry);
+            CacheEntriesByRequest.Add(correlationId, cacheEntry);
+        }
+
+        [Fact]
+        public virtual void New_Entries_Should_Be_Added_With_Individual_Entry_Options()
+        {
+            PendingRequests.ForEach(p => AddCreateEntryExpectation(p.Content.CorrelationId));
+            PendingRequests.ForEach(CorrelationManager.AddPendingRequest);
+
+            Cache.Received(PendingRequests.Count).CreateEntry(
+                Arg.Is<object>(o => ((ByteString) o).ToCorrelationId().Id != Guid.Empty));
+
+            var createEntryCalls = Cache.ReceivedCalls()
+               .Where(ci => ci.GetMethodInfo().Name == nameof(IMemoryCache.CreateEntry));
+
+            createEntryCalls.Select(c => c.GetArguments()[0]).Cast<ByteString>()
+               .Should().BeEquivalentTo(PendingRequests.Select(p => p.Content.CorrelationId));
+
+            foreach (var cacheEntry in CacheEntriesByRequest.Values)
+            {
+                cacheEntry.ExpirationTokens.Count.Should().Be(1);
+                cacheEntry.PostEvictionCallbacks.Count.Should().Be(1);
+            }
+
+            CheckExpirationTokensAreDifferentForEachEntry();
+
+            CheckCacheEntriesCallback();
+        }
+
+        private void CheckExpirationTokensAreDifferentForEachEntry()
+        {
+            for (var i = 0; i < CacheEntriesByRequest.Count; i++)
+            for (var j = i + 1; j < CacheEntriesByRequest.Count; j++)
+            {
+                CacheEntriesByRequest[PendingRequests[i].Content.CorrelationId]
+                   .Should().NotBeSameAs(CacheEntriesByRequest[PendingRequests[j].Content.CorrelationId]);
+            }
+        }
+
+        protected abstract void CheckCacheEntriesCallback();
+
+        protected void FireEvictionCallBackByCorrelationId(ByteString correlationId)
+        {
+            var pendingRequest = PendingRequests.Single(p => p.Content.CorrelationId == correlationId);
+            CacheEntriesByRequest[correlationId].PostEvictionCallbacks[0].EvictionCallback
+               .Invoke(null, pendingRequest, EvictionReason.Expired, null);
+        }
+
+#pragma warning disable 693
+        protected void TryMatchResponseAsync_Should_Match_Existing_Records_With_Matching_Correlation_Id<T>()
+#pragma warning restore 693
+            where T : IMessage, new()
+        {
+            var responseMatchingIndex1 = new T().ToProtocolMessage(
+                PeerIds[1],
+                PendingRequests[1].Content.CorrelationId.ToCorrelationId());
+
+            var request = CorrelationManager.TryMatchResponse(responseMatchingIndex1);
+            request.Should().BeTrue();
+        }
+
+#pragma warning disable 693
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Do not suppress issues.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4318201)